Content title numbering

vidhyasagargm opened the pull request:
see #1 
created a new branch and implemented the numbering changes to it.

marius-meissner commented on widgets/.WikiRichText.php.swp:
Please avoid checking in .swp files. Maybe add them to .gitignore.

vidhyasagargm replied:
Yeah i will follow that next time I commit.

marius-meissner commented on widgets/PageListItemTitle.php:
```diff
@@ -103,6 +105,25 @@ public function run()
             'level' => $this->level,
         ]);
     }
+    protected function generateNumbering($level)
```
```suggestion

    protected function generateNumbering($level)
```

marius-meissner commented on widgets/PageListItemTitle.php:
```diff
@@ -103,6 +105,25 @@ public function run()
         ]);
     }
+    protected function generateNumbering($level)
+    {
```
Please add unit tests for logic. See for general reference: https://docs.humhub.org/docs/develop/testing/

vidhyasagargm replied:
UPDATE: I have implement the unit test logic for generateNumbering and commited it below

marius-meissner reviewed (changes requested):
Thanks for the PR :+1: Good start, but a few improvements should still be implemented.

PR has been deployed to staging system: https://humhub.staging.neoatlas.org/
@TobiNeo Please check from your side and add suggestions to this PR.

Suggestions for changes from the user's point of view (acceptance criteria):
- [ ] Add a option toggle for enabling/disabling content number to admin panel
- [ ] Content summary contains double the numbering:
![image](https://github.com/user-attachments/assets/01880beb-9739-4260-8d6d-e5756194e616)

vidhyasagargm commented:
ERROR: I pushed an unwanted commit which i have reverted back. Sorry for the inconvience

vidhyasagargm commented:
UPDATE: Implemented toggle for pagelist numbering with unit testing.

vidhyasagargm commented:
UPDATE:
Fixed the double numebring in content and body. Added toggle button. The toggle button used for this is the same logic as for the overvie numbering hence i didnt implement unit testing for the logic again. The numbering for the body was done purely using CSS styling. 
This is how the numbering looks right now(The  numbering on the right side was for reference):
![Screenshot (18)](https://github.com/user-attachments/assets/185881fc-9194-478d-a418-2b7f930a9e89)

TobiNeo commented:
Hi Vidhyasagar,
thanks for your efforts so far. Some comments on the more general questions:

> * I have implemented the toggle button only in english version, which all languages should i implement it in?

Please also implement it in German, you may call it "Nummerierung umschalten". All other languages are not needed for now.



> UPDATE: Fixed the double numebring in content and body. Added toggle button. The toggle button used for this is the same logic as for the overvie numbering hence i didnt implement unit testing for the logic again. The numbering for the body was done purely using CSS styling. This is how the numbering looks right now(The numbering on the right side was for reference): ![Screenshot (18)](https://private-user-images.githubusercontent.com/85664242/369443462-185881fc-9194-478d-a418-2b7f930a9e89.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcwOTgzOTIsIm5iZiI6MTcyNzA5ODA5MiwicGF0aCI6Ii84NTY2NDI0Mi8zNjk0NDM0NjItMTg1ODgxZmMtOTE5NC00NzhkLWE0MTgtMmI3ZjkzMGE5ZTg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIzVDEzMjgxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiZGM3ZTY1YTlmYWEzZDk4YWE5YThmZjNmYWE0NzY1OGFlYjQ4NDI4NDQ1NmIzYTNiY2YwNWM0YmI5ZTk0ODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.O8TiV_gmdzgc8tyGd7xhBsT4AJCBX2MdJewuGG7RJeE)

Looks good, at least from a non-coding point of view :)

Best regards
Tobias

vidhyasagargm commented:
UPDATE:
The numbering button appeared on the edit page of wiki. This issue was resolved.

marius-meissner reviewed (changes requested):
Looks much better now. Very nice idea with the CSS solution :+1: 

Changes have been deployed to staging system. 
Upfront, one suggestions from user level.

**Preserve setting**
Currently, the number setting is not retained when you switch to a subpage. 
But it would also be nice if the setting could be saved permanently at user level. Preferably on the server side, so that the setting is retained even if you change workstation or device.

vidhyasagargm commented:
INFO: 
The CSS styling doesnt change just by refreshing the page. We have to go to Administration>settings>advanced and click on clear cache to display the changes.

vidhyasagargm commented:
Implemented preserve settings on the server side. I couldn't verify it on cross platform but i did try logging in with an incognito window and the toggle numbering was preserved. I created two different settings one for Overview numbering and one for wiki page numbering. There are two different unit test file for the two different numbering settings. both have similar logic.

vidhyasagargm commented:
UPDATE:
Implemented Preserve Settings that stores the user settings on the server end. 

@marius-meissner Changes are ready for review.

vidhyasagargm commented:
ISSUE:

- [ ] Missing numbering when a subheading doesnt exist for example 2.0.1

marius-meissner commented:
> could you give me the correct structure required for numbering so that i can change it in the TOC as well as the content.

In my view, it is sufficient to simply follow the current TOC logic.

vidhyasagargm commented:
Hello @marius-meissner ,
I have implemented the same numbering of TOC  and Content. I had to modify the Javascript and chnage the asset file. With codeception unit testing i can only test php file. How do i perform unit testing for this javascript changes?
Apart from unit testing the code is ready for review.

marius-meissner commented:
> How do i perform unit testing for this javascript changes?

If there is no possibility for JS UnitTests in the Upstream HumHub or Wiki, it is currently fine that we do not implement any there.

marius-meissner commented:
Furthermore please add a setting (e.g. checkmark) to [wiki/config](https://humhub.staging.neoatlas.org/wiki/config) whether content numbering should be enabled or disabled by default for all users.

vidhyasagargm commented:
UPDATE:
Grunt build was used to generate min, bundle and less files.

vidhyasagargm commented on resources/js/humhub.wiki.bundle.js:
```diff
@@ -199,6 +200,49 @@ humhub.module('wiki.Page', function (module, require, $) {
+                $header.prepend('<span class="wiki-header-numbering">' + numberString + ' </span>');
+            });
+        }
+    };
```
> 2. Logs of duplicate logic exist between `Page.prototype.buildIndex` and your method. So please try to abstract the shared logic. Maybe add a third method which does the parsing of the headers, which accepts a labmda action, which is then executed for each header item.

instead of creating another method i just appened it into the buildindex funciton with a condiiton to inject the numbering into the headings

vidhyasagargm commented:
UPDATE:
Added setting to [wiki/config](https://humhub.staging.neoatlas.org/wiki/config). 

@marius-meissner the code is ready for review

vidhyasagargm commented:
UPDATE: 
Removed empty spaces and small naming corrections apart from that code ready for review

marius-meissner commented on messages/de/base.php:
```diff
@@ -86,4 +86,6 @@
     'Hide Navigation Entries of this module globally by default' => '',
     'Hide Navigation Entry' => '',
     'No wiki pages found.' => '',
+    'Enable Numbering' => 'Nummerierung umschalten',
+    'Disable Numbering' => 'Nummerierung abschaltbar',
```
```suggestion
    'Enable Numbering' => 'Nummerierung aktivieren',
    'Disable Numbering' => 'Nummerierung deaktivieren',
```

marius-meissner commented on resources/css/humhub.wiki.min.css:
```diff
@@ -1 +1 @@
-.layout-content-container .wiki-welcome h1{font-size:36px;font-weight:700}.layout-content-container .wiki-welcome h2{font-size:14px;font-weight:200;line-height:24px;margin:10px 100px}@media (max-width:768px){.layout-content-container .wiki-welcome h2{margin:10px}}.layout-content-container .wiki-bg{padding:150px 0 150px;background-repeat:no-repeat;background-position:center 100px}@media only scree ...
\ No newline at end of file
+.layout-content-container .wiki-welcome h1{font-size:36px;font-weight:700}.layout-content-container .wiki-welcome h2{font-size:14px;font-weight:200;line-height:24px;margin:10px 100px}@media (max-width:768px){.layout-content-container .wiki-welcome h2{margin:10px}}.layout-content-container .wiki-bg{padding:150px 0 150px;background-repeat:no-repeat;background-position:center 100px}@media only scree ...
```
Please don't commit minimized/compiled files.
It is true that these are contained directly in GIT, but as far as I know this is done automatically by the build system before a release. Otherwise it leads to many problems, as there will be constant conflicts.

marius-meissner commented on resources/js/humhub.wiki.Page.js:
```diff
@@ -134,6 +134,11 @@ humhub.module('wiki.Page', function (module, require, $) {
+
+            // Added numbering to content
+            if(document.querySelector('.numbered')){
+                $(this).prepend('<span class="wiki-header-numbering">' + numberString + '</span>');
```
Please add a blank between the heading and the index.
```suggestion
                $(this).prepend('<span class="wiki-header-numbering">' + numberString + ' </span>');
```

marius-meissner replied:
As just discussed with @JuergStae it even looks more structured if a space in combination with dot is used:
```suggestion
                $(this).prepend('<span class="wiki-header-numbering">' + numberString + '. </span>');
```

marius-meissner commented on resources/css/humhub.wiki.css:
```diff
@@ -787,3 +787,8 @@
     padding: 0;
   }
 }
+/* CSS Styling for toggle button */
+.toggle-numbering {
+  margin-left: auto;
+  padding-right: 10px;
+}
```
Same here, file will be generated by less build system.
```suggestion
```

marius-meissner commented on resources/css/humhub.wiki.less:
```diff
@@ -1056,3 +1056,5 @@
     padding: 0;
   }
 }
+/* CSS Styling for toggle button */
+.toggle-numbering {margin-left: auto;padding-right: 10px;}
```
Please stick to code style.

```suggestion
/* CSS Styling for toggle button */
.toggle-numbering {
  margin-left: auto;
  padding-right: 10px;
}
```

marius-meissner commented on resources/js/humhub.wiki.bundle.min.js:
```diff
@@ -1 +1 @@
-humhub.module("wiki",function(i,t,s){function e(t,e,i){c.push({$node:t,$trigger:e,condition:i})}var n=t("ui.widget").Widget,r=t("event"),l=t("ui.view"),a=t("client"),o=t("ui.loader"),c=[],t=n.extend();t.prototype.init=function(){s(window).off("scroll.wiki").on("scroll.wiki",function(){s.each(c,function(t,e){var i=s(window).scrollTop()+l.getContentTop()>e.$trigger.offset().top,n=!e.condition||e.co ...
\ No newline at end of file
+humhub.module("wiki",function(i,t,s){function e(t,e,i){c.push({$node:t,$trigger:e,condition:i})}var n=t("ui.widget").Widget,r=t("event"),l=t("ui.view"),a=t("client"),o=t("ui.loader"),c=[],t=n.extend();t.prototype.init=function(){s(window).off("scroll.wiki").on("scroll.wiki",function(){s.each(c,function(t,e){var i=s(window).scrollTop()+l.getContentTop()>e.$trigger.offset().top,n=!e.condition||e.co ...
```
Same here, please do not check in generated files.

marius-meissner commented on resources/js/humhub.wiki.bundle.js:
```diff
@@ -272,6 +272,11 @@ humhub.module('wiki.Page', function (module, require, $) {
             });
             $li.append($anchor);
             $list.append($li);
+
+            // Added numbering to content
+            if(document.querySelector('.numbered')){
+                $(this).prepend('<span class="wiki-header-numbering">' + numberString + '</span>');
+            }
```
Same here, please do not check in generated files.

```suggestion
```

marius-meissner commented on resources/js/humhub.wiki.Page.js:
```diff
@@ -134,6 +134,11 @@ humhub.module('wiki.Page', function (module, require, $) {
             $list.append($li);
+
+            // Added numbering to content
+            if(document.querySelector('.numbered')){
```
Please use strict type checks with JS.

```suggestion
            if(document.querySelector('.numbered') !== null){
```

marius-meissner commented on tests/codeception/functional/NumberingButtonCest.php:
```diff
@@ -0,0 +1,73 @@
+        $numberingEnabled = $module->settings->contentContainer($user)->get('wikiNumberingEnabled');
+
+        $I->seeElement('a.toggle-numbering');
+        if ($numberingEnabled) {
+            $I->see('Disable Numbering', 'a.toggle-numbering');
+        }   
+        else {
+            $I->see('Enable Numbering', 'a.toggle-numbering');
+        } 
```
The test already looks very good. Well done.
However, IF statements should be avoided in unit tests. It would be better to ensure a fixed condition at the beginning of the test (e.g. numbering switched off), then trigger an action and check for a fixed expected value.

However, you can execute several actions in succession in one test, e.g.
1. start condition: Numbering switched off
2. action: press button
3. check whether numbering is now switched on
4. action: press button
5. check whether numbering is now switched off

marius-meissner reviewed (changes requested):
Thanks! Looks a lot better now. :+1: 

Changes have been deployed to staging system, but there is no config item visible at https://humhub.staging.neoatlas.org/wiki/config.
Please check whether something needs to be done to our deployment commands or on the code side.

vidhyasagargm commented:
> Changes have been deployed to staging system, but there is no config item visible at https://humhub.staging.neoatlas.org/wiki/config. Please check whether something needs to be done to our deployment commands or on the code side.

Yeah thats more on the coding side i didnt put a checkbox on the index page of config

vidhyasagargm commented:
Hello @marius-meissner ,

- I have updated all the changes, now it should be visible on wiki/config and i have removed all the generated files.
- I saw  the wiki documentation on deployment command I believe there is a typo on the remove command.

vidhyasagargm commented:
UPDATE: 
I used DeepL for the translation of the content of config page.

marius-meissner reviewed (changes requested):
Looks wonderful now. Just some tiny optimizations and then we are ready for production from my side.

vidhyasagargm commented on tests/codeception/functional/NumberingButtonCest.php:
```diff
@@ -0,0 +1,43 @@
+use wiki\FunctionalTester;
+use Yii;
+
+class NumberingButtonCest
```
Hi @marius-meissner ,
For functional test and acceptance test the Class name and file name has to end with Cest or else the test won't run in codeception. This is as given in the Codeception [documentation](https://codeception.com/docs/AdvancedUsage).

vidhyasagargm commented:
Hello @marius-meissner,
Thanks a lot for your feedback
UPDATE:
- The suggested changes has been done except for the test file 
- The code is ready for review

marius-meissner reviewed (changes requested):
Alright, looks almost perfect now.
I just noticed one little thing. When resorting the pages, the index should get updated.
[move_page.webm](https://github.com/user-attachments/assets/67110f69-30ca-4ab7-a420-3a12025c77d7)


Maybe you can somehow trigger a (partial) refresh when the sorting request is done:
https://github.com/Neo-Atlas/humhub_wiki/blob/54f20c73b6115298b3d93c16e0d37cdcdf681aaa/resources/js/humhub.wiki.CategoryListView.js#L257

vidhyasagargm commented:
Hello @marius-meissner 
UPDATE:
Partial reload has been done as Pjax is default enabled just doing client.reload would execute a partial reload on the components, which visiually is faster than whole page reload.

The code is ready for review

vidhyasagargm commented:
UPDATE:
Added condition for the partial reload so that it does'nt reload when there is no numbering.

The code is ready for review.

marius-meissner reviewed (approved):
Looks good from our side.
PR will be merged after final feedback of next JF meeting.

In the meanwhile, please create an upstream PR in https://github.com/humhub/wiki:

Please add an explanation what the changed is about
Add a link to this PR as reference, mentioning that an internal code review process has been already done upfront.

vidhyasagargm commented:
Greetings @marius-meissner,
The upstream [PR]((https://github.com/humhub/wiki/pull/357)) has been reviewed. There are some Fix and Enh recommended. Some of it is cause of not using grunt build for CSS and JS files.
